sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1008,6 +1088,27 @@ bool _pi_queue::isDiscardEvents() const {
   return ((this->Properties & PI_EXT_ONEAPI_QUEUE_DISCARD_EVENTS) != 0);
 }
 
+void _pi_queue::setLastCommandEvent(pi_event Event) {
+  //  Save copy of the last event. We can put event into the cache only if it is
+  //  not referenced externally anymore.
+  if (Event &&
+      supportsInOrderQueueOptimization(Event->isHostVisible(),
+                                       Event->IsDiscarded) &&
+      !Event->hasExternalRefs()) {
+    CandidateForReuse =
+        new _pi_event(Event->ZeEvent, Event->ZeEventPool, Context,
```

Comment:
> where is this one going to be deleted?
Note. I will try to get rid of CandidateForReuse and use LastCommandEvent only per comment below.

pi_event objects get deleted in piEventRelease, we reuse only native handles. pi_event objects are only wrappers which get created and deleted in piEventRelease while ze_event handles are reused.

> why do you need a to allocate a whole new PI event?
We need to allocate a whole pi_event when caching in the queue level. pi_event object encapsulates data associated with a particular command (kernel, mapped buffer, list of dependent events, etc), we need to take some actions on this data when the event is *completed* (release the kernel, free allocated memory, cleanup list of dependent events appropriately). 

But as you've summarized in this comment https://github.com/intel/llvm/pull/6643#discussion_r957369974
we reuse native event handles before pi_event is completed (for in-order queue we have more info about dependency between commands). We can't reuse whole pi_event object before its completion and we need to re-create pi_event using same native handles. 

Example, 

1. kernelA -> pi_event1 {ze_event1, ptr to kernelA, deplist1}
2. kernelB-> pi_event2 {ze_event2, ptr to kernelB, deplist2}
3. reset ze_event1  -> at this point pi_event1 is not completed yet, we can't reuse whole pi_event1, because we still going to need info in pi_event1 like event list - deplist1. So we create another object pi_event3 with the same native handle ze_event1. 
4. kernelC-> pi_event3 {ze_event1, ptr to kernelC, deplist3}

As you see pi_event1 and pi_event3 have same native handle - ze_event1 but different data because they are associated with different commands and we need to keep all this info because neither pi_event1 nor pi_event3 is not completed yet and we are going to need this info when they are completed.